### PR TITLE
Expand mouse capture and release logic

### PR DESCRIPTION
**Pull request by krcroft**
This commit:
- replaces the existing `true` and `false` mouse capture configuration with three choices:
  - **onstart**, which captures the mouse immediately on starting DOSBox. You now don't need to mouse-around and click once just to get started: you're mouse is active from the get-go, which matches original DOS behavior.
  - **onclick**, which captures the mouse after manually clicking into the DOSBox window. This is equivalent to the existing DOSBox behavior, and is the default. 
  - **never**, which disables capturing the mouse in windowed-mode. The mouse cannot be accidentally captured and is always useable for other tasks.  This is perhaps useful for non-mouse games.
  - In all three modes the mouse is captured in fullscreen because at that point DOSBox has exclusive focus so you might as well have control of the mouse.
- The capture-state is retained when exiting fullscreen (for `onclick` and `onstart`) so you don't have to _"click again"_ to regain mouse control to continue playing.  Where as `never` always releases the mouse as promised in window-mode. 
- Renames the conf file setting to `capturemouse` to better describe _when the mouse will be captured_: For example, it now says what it does: `capturemouse = onstart`, or `capturemouse = onclick`.
- Previously the code used four separate state-tracking 'bool's to control the mouse capture behavior; this commit boils them down to one.
- Add a boolean config option `middlerelease`, (true/false), allowing the middle-mouse button to release the captured mouse, in window-mode.
- A one-line message is now printed to the console indicating when the mouse is captured and released.
- All of the above behavior is commented into the conf file when it's written (6 lines of text).

### CODE REVIEW

**Comment by krcroft** [comment]
Will add another commit that tidies up `if`, `for` and `while` spacing.

**Comment by krcroft** [comment]
I didn't mention it in the commit, but the prior config name of 'autolock' never did anything automatically.  Even when enabled,  user had to manually navigate the cursor inside the DOSBox window and manually click to capture the mouse.

**Comment by krcroft** [comment]
Vogons user **citrixscu** [mentioned](https://www.vogons.org/viewtopic.php?p=820747#p820747): 

> Seems to work ok on my Catalina laptop.

**Comment by krcroft** [comment]
More good news from @citrixscu:

> capturemouse=onstart --> works as intended
> capturemouse=onclick --> works as intended
> capturemouse=never --> works as intended

**Comment by krcroft** [comment]
Renamed PR to include the word 'correct' because baseline does not properly release on MiddleClick. This PR fixes that behavior, now tested working on:
 - Linux, real hardware
 - MacOS VM
 - MacOS real hardware
 - Windows 7 and 10 VM
 - Windows 10 hardware

File: src/gui/sdlmain.cpp [hunk 1]
```diff
@@ -846,28 +849,21 @@ Bitu GFX_SetSize(Bitu width,Bitu height,Bitu flags,double scalex,double scaley,G
...
-		SDL_SetRelativeMouseMode(SDL_FALSE);
-		if (sdl.mouse.autoenable || !sdl.mouse.autolock) SDL_ShowCursor(SDL_ENABLE);
-	}
-        mouselocked=sdl.mouse.locked;
+SDL_bool mouse_is_captured = SDL_FALSE; // global for mapper
+void GFX_ToggleMouseCapture(void) {
+	SDL_ShowCursor(mouse_is_captured);
+	mouse_is_captured = mouse_is_captured ? SDL_FALSE : SDL_TRUE;
+	SDL_SetRelativeMouseMode(mouse_is_captured);
+	LOG_MSG("SDL: mouse is %s", mouse_is_captured ? "captured" : "released");
```
Comment (dreamer) [inline src/gui/sdlmain.cpp]: It seems like `SDL_ShowCursor` changed from SDL 1.2 to SDL 2.0 in a subtle, (backwards compatible) way. This is good opportunity to use it according to [documentation](https://wiki.libsdl.org/SDL_ShowCursor); `SDL_ShowCursor` now accepts int and recognizes one of three values: SDL_QUERY, SDL_DISABLE, or SDL_ENABLE. HOWEVER - perhaps we don't need to invoke this function at all.

[`int SDL_SetRelativeMouseMode(SDL_bool)`](https://wiki.libsdl.org/SDL_SetRelativeMouseMode) can fail, so if it does, let's log the failure and error message and restore our internal value of `mouse_is_captured` field. I noticed in "Remarks" section:

> While the mouse is in relative mode, the cursor is hidden (…)

:) 

If it does indeed work, then I think it warrants a comment in code.

File: src/gui/sdlmain.cpp [hunk 2]
```diff
@@ -1461,9 +1453,36 @@ static void GUI_StartUp(Section * sec) {
...
+		// If an invalid value is provided, the section handler
+		// should have automatically set the default to 'onclick'
+		assert(sdl.mouse.capture_choice == MouseCaptureType::OnClick);
+	}
+	// Capture the mouse straight away ...
+	if (sdl.desktop.fullscreen || sdl.mouse.capture_choice == MouseCaptureType::OnStart) {
+		GFX_ToggleMouseCapture();
+	// Otherwise simply entire the cursor is visible
+	} else {
+		SDL_ShowCursor(1);
```
Comment (dreamer) [inline src/gui/sdlmain.cpp]: Oh, `SDL_ShowCursor` usage. If we need this, then use `SDL_ShowCursor(SDL_ENABLE)`.

File: src/gui/sdlmain.cpp [hunk 3]
```diff
@@ -1473,40 +1492,28 @@ static void GUI_StartUp(Section * sec) {
...
+						  mouse_is_captured);
 }
 
 static void HandleMouseButton(SDL_MouseButtonEvent * button) {
 	switch (button->state) {
 	case SDL_PRESSED:
-		if (sdl.mouse.requestlock && !sdl.mouse.locked) {
-			GFX_CaptureMouse();
+		if ((sdl.mouse.capture_choice != MouseCaptureType::Never && !mouse_is_captured)
+		    || (button->button == SDL_BUTTON_MIDDLE && mouse_is_captured)) {
```
Comment (dreamer) [inline src/gui/sdlmain.cpp]: I'm afraid some games might use middle mouse button… I thought Settlers is the one, but that one uses left+right "special" click: d88f6db3b9193d276756b01703c6e9822e8646e5. I googled a bit and the only one I could find was a [user-created mod for Dune II](https://forum.dune2k.com/topic/27527-dune-ii-mouse-helper-utility-for-improving-dos-version-controls/), but I think we need to do some digging.

And I believe macOS does not support middle mouse button out of the box and requires a separate app installed? (form over function, I guess)

File: src/gui/sdlmain.cpp (hunk 1 above)
Comment (krcroft) [reply src/gui/sdlmain.cpp]: Great find; and it works - both hiding and revealing the cursor on each toggle. 
Comment and error check added.

File: src/gui/sdlmain.cpp (hunk 2 above)
Comment (krcroft) [reply src/gui/sdlmain.cpp]: Another nice improvement to make the intent clear. 
Updated.

File: src/gui/sdlmain.cpp (hunk 3 above)
Comment (krcroft) [reply src/gui/sdlmain.cpp]: MacOS honors middle click (and wheel) if the USB device provides it.  Just need to plug in a PC mouse.  Behaviour is also confirmed on a VM that provides 3-button passthrough (like VirtualBox).

But yeah - I didn't realize DOS games recognized the middle mouse button! By the time I got a 3-button mouse, I had moved onto Windows 95.  Turns our the early 90s Sierra VGA point and click games would rotate the action cursor (walk/talk/look/...) on a middle press.  Who know!  

So I've made this configurable and updated the PR message. 

Edit: tested on all three platforms now.

**Comment by dreamer** [comment]
As for the code itself, I see nothing serious - just some aesthetic stuff - I'll mention it in next round of review.

I tested the patch a little bit, and ran into some troubles - I use Wayland with dosbox running via XWayland, so this might be causing the problem; I'll do some more testing tomorrow. The troubling behaviour I'm seeing is:

With `onstart` mouse is not captured at all if the cursor is outside of the window - it is caught only after cursor moves over the dosbox window (and from that point forward it's ok).

With `never` mouse seems to be completely disabled (tested Windows 3.11), I can't move the cursor at all; previously the cursor was moving with bad acceleration, but it was moving (Windows 3.11 has notable issues with cursor, so I need to repeat this test with some games).

Let's not rush with fixes until we know the root cause - perhaps we need to report some issues to SDL2 about Wayland or XWayland support.

**Comment by krcroft** [comment]
> With onstart mouse is not captured at all if the cursor is outside of the window - it is caught only after cursor moves over the dosbox window (and from that point forward it's ok).

Under baseline DOSBox, move your cursor outside of the window and then press Ctrl-F10.  Does that lock it / and hide the cursor?  Both Ctrl-F10 and onstart just call the same function; however the subtle difference is that, in order to press Ctrl-F10, you need to click and give DOSBox active focus.  Where as with onstart, DOSBox might not even be focussed (maybe we need to force it to focus to foreground on startup).
 
If you build 6737c28b66c99c439637672efc1a66ceafd57ce6  , does that behave differently? It still contains `SDL_ShowCursor(...)` that explicitly hides/shows the cursor on lock events. 

> With `never`, mouse seems to be completely disabled (tested Windows 3.11), I can't move the cursor at all; previously the cursor was moving with bad acceleration, but it was moving (Windows 3.11 has notable issues with cursor, so I need to repeat this test with some games).

`never` should behave like this: your mouse is uncapturable by DOSBox when in window-mode.  Clicking inside DOSBox and Ctrl-F10 are disabled, so there's no way to capture the mouse or use it when DOSBox is windowed. But if you switch to fullscreen, your mouse should become active and fully useable inside DOSBox.  That how it behaves on X11, MacOS, and Windows.  Toggling back out of fullscreen will again uncapture your mouse (and leave it uncapturable).

**Comment by krcroft** [comment]
@dreamer , the latest commit focuses the window using SDL_RaiseWindow() prior to capturing (via setting MouseRelativeMode) when `onstart`.  Curious if that helps Wayland.  If it doesn't, then I think we have a solid issue we can send to SDL and see what they think.

**Comment by dreamer** [comment]
Putting the problems I had with `onstart` aside for a while.

This is the problem I have: https://imgur.com/a/hiQQeXf (sorry about imgur link; GitHub does not allow me to post WebM and GIF does not capture cursor movement well enough).

I am not saying, that `never` should be changed. I think we need an option that provides the same behaviour as old `autolock=false`, as it provides very intuitive behaviour for many games (and practically all non-games software, but games are deciding factor).

Maybe `capturemouse=never` should trigger old `autolock=false` behaviour and there should be fourth option (something like `capturemouse=disable_mouse`?).

**Comment by krcroft** [comment]
Thanks @dreamer , that visual helped a lot; I was able to reproduce it. 
I see a couple dimensions:
- **_When_** is the mouse captured? onstart, onclick, or never 
- **_How_** is the mouse captured?  bounded inside DOSBox or seamless with the desktop
- **_If_** the middle-mouse button can also be used for release

Thinking about how best to make this configurable, we could go with combined or separation config options. 

```
captureevent = onstart, onclick, never
capturetype = bounded, seamless
middlerelease = true, false
```

Or combined, some examples:

```
mousecapture = never
mousecapture = onclick bounded
mousecapture = onstart seamless
mousecapture = onstart bounded middlerelease
```

The combined one feels like the tidiest inside the conf file, and with a good description I think users would be comfortable with it (without being confusing).

What do you think?

**Comment by dreamer** [comment]
I like the combined solution a lot - it's very explicit, option names are self-documenting and it provides more user control than upstream DOSBox.

A note: when implementing SDL2 I renamed `renderer` to `texture_renderer`, which now I see is the only key name using `_`; do you prefer `mousecapture` or maybe `mouse_capture` or `capture_mouse` (all these options are fine with me).

**Comment by krcroft** [comment]
Yes; I also felt compelled to conform to the existing `squashedandlower` convention, but it terrible.

`capture_mouse onstart ...` reads the most natural to me.  We can also adjust these labels before the merge, if anyone else wants to toss out some suggestions.

**Comment by krcroft** [comment]
Closing for now. Will rework this PR into just a logic commit, and cut out all the reformatting, which is currently blocking me from squashing the commits around it.

**Pull request closed without merging**